File: C#/Minimum Depth of Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution111 {
    public int MinDepth(TreeNode root) {
        
        /// DFS
        if (root == null)	return 0;
        if (root.left == null)	return MinDepth(root.right) + 1;
        if (root.right == null) return MinDepth(root.left) + 1;
        return Math.Min(MinDepth(root.left),MinDepth(root.right)) + 1;
        
    }
}


public class Solution111 {
    public int MinDepth(TreeNode root) {
        
        ///BFS
        if (root == null) return 0;
        
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        
        int depth = 0;
        while (queue.Count != 0)
        {
            int size = queue.Count;
            depth++;
            for (int i = 0; i < size; ++i)
            {
                var curr = queue.Dequeue();
                
                if (curr.left == null && curr.right == null)
                {
                    return depth;
                }
                
                if (curr.left != null) queue.Enqueue(curr.left);
                
                if (curr.right != null) queue.Enqueue(curr.right);
            }
        }
        return depth;
        
    }
}

File: Golang/Minimum Depth of Binary Tree.go
//BFS
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	q := []*TreeNode{root}
	out := 1
	for len(q) != 0 {
		former := len(q)
		for i := 0; i < former; i++ {
			if q[i].Left == nil && q[i].Right == nil {
				return out
			}
			if q[i].Left != nil {
				q = append(q, q[i].Left)
			}
			if q[i].Right != nil {
				q = append(q, q[i].Right)
			}
		}
		q = q[former:]
		out++
	}
	return -1
}

//DFS
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left, right := minDepth(root.Left), minDepth(root.Right)
	if left == 0 || right == 0 {
		return left + right + 1
	}
	return min(left, right) + 1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}